### PR TITLE
Basic Smarty templating

### LINKED ISSUE #1820: Add basic smarty templating support
I am using `js-beautify` to format[ `smarty` ](https://www.smarty.net/) files. It would be helpful if I can extend/modify the `Tokenizer` prototype. I've noticed that the module is not exported. Is there some other way to do it?

I am seeking for an API to extend the `Tokenizer` functions.

### PR DESCRIPTION
**Pull request by aswinkumar863**
# Description
- [x] Source branch in your fork has meaningful name (not `master`)

Fixes Issue: 

Added basic [smarty ](https://www.smarty.net/) templating support. Resolves #1820

# Before Merge Checklist 
These items can be completed after PR is created.

(Check any items that are not applicable (NA) for this PR)

- [x] JavaScript implementation
- [x] Python implementation (NA if HTML beautifier)
- [ ] Added Tests to data file(s)
- [x] Added command-line option(s) (NA if
- [x] README.md documents new feature/option(s)

### CODE REVIEW

**Comment by aswinkumar863** [comment]
> It looks like smarty templating is incompatible with django. Specifically, I think `.until_after(r"}")` will result in django templates closing too soon.

I am not sure how to overcome this. And I hope _django_ `(python)` and _smarty_ `(php)` will not be used at the same time.

**Comment by bitwiseman** [comment]
@aswinkumar863 
could you fix the formatting failures?

**Comment by aswinkumar863** [comment]
@bitwiseman I hope you are asking about the failures that I mentioned [here](https://github.com/beautify-web/js-beautify/issues/1820#issuecomment-714210759). After I turned off the [option](https://github.com/aswinkumar863/js-beautify/commit/15747529f5fc29889ebc53ba6badbd9b956e103a#r44703790) in https://github.com/beautify-web/js-beautify/pull/1861/commits/3fb86dce39e20c9ab767c2a33785712505416686, **all test cases are passed**.

**Comment by bitwiseman** [comment]
@aswinkumar863 
No, you need to run `make` and then commit the changes (autoformatting).

**Comment by bitwiseman** [comment]
@aswinkumar863  
I think you will need to have a check if `smarty` is enabled, and disables `django` and `handlebars`.   So, `auto` does not include `smarty`.   

Then you can add tests to verify smarty templating works on its own.

**Comment by bitwiseman** [comment]
Ah the formatting failure was due to an external tool change.

**Comment by bitwiseman** [comment]
@aswinkumar863 
Okay, now that the build is fixed you can safely do the work discussed above and adding tests.

**Comment by aswinkumar863** [comment]
> @aswinkumar863
> Okay, now that the build is fixed you can safely do the work discussed above.

@bitwiseman Sure, I will work as discussed.

**Comment by bitwiseman** [comment]
> I never wrote a test case in my life my bad. I need to learn some basics and do that. I will reach you out if I face any problem.

Make changes to https://github.com/beautify-web/js-beautify/blob/master/test/data/html/tests.js and then run `make js`.

You can specify `input` and expected `output` to compare against, or you can specify `unchanged` if they should be the same. 

I made a couple tweaks and added smarty to the basic templating matrix.

**Comment by bitwiseman** [comment]
@aswinkumar863 When do you think you'll have time to finish this up?  It would be nice to get it into the next release.

**Comment by aswinkumar863** [comment]
> @aswinkumar863 When do you think you'll have time to finish this up? It would be nice to get it into the next release.

I will try to finish within this week.

File: js/src/core/templatablepattern.js [hunk 1]
```diff
@@ -63,8 +63,9 @@ function TemplatablePattern(input_scanner, parent) {
     django: pattern.starting_with(/{%/).until_after(/%}/),
     django_value: pattern.starting_with(/{{/).until_after(/}}/),
     django_comment: pattern.starting_with(/{#/).until_after(/#}/),
-    smarty: pattern.starting_with(/{(?=[^}\s\n])/).until_after(/}/),
-    smarty_comment: pattern.starting_with(/{\*/).until_after(/\*}/)
+    smarty: pattern.starting_with(/{(?=[^}{\s\n])/).until_after(/[^\s\n]}/),
+    smarty_comment: pattern.starting_with(/{\*/).until_after(/\*}/),
+    smarty_literal: pattern.starting_with(/{literal}/).until_after(/{\/literal}/)
```
Comment (aswinkumar863) [inline js/src/core/templatablepattern.js:68]: According to docs, literals are used around Javascript or stylesheet. So allowing formatting inside literal would be better.

**Comment by aswinkumar863** [comment]
@bitwiseman I have added a few tests in b0dc440d2c1ee45b4052d886d0cc6032c440b921. After finalizing https://github.com/beautify-web/js-beautify/pull/1861#discussion_r548602498, you can proceed to release.

File: js/src/core/templatablepattern.js (hunk 1 above)
Comment (bitwiseman) [reply js/src/core/templatablepattern.js:68]: Yes, however this is basic support.  Feel free to file an issue for that and do the work in a future PR.

Comment (aswinkumar863) [reply js/src/core/templatablepattern.js:68]: Okay, let's do that later.

**Review by bitwiseman — APPROVED** [review]

**Pull request merged**